src/analysis/raster/qgskde.cpp

```diff
@@ -164,13 +176,13 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::addFeature( const
     unsigned int xPosition = ( ( ( *pointIt ).x() - mBounds.xMinimum() ) / mPixelSize ) - buffer;
     unsigned int yPosition = ( ( ( *pointIt ).y() - mBounds.yMinimum() ) / mPixelSize ) - buffer;
 
+    // calculate buffer around pixel
+    QgsRectangle extent( ( *pointIt ).x() - radius, ( *pointIt ).y() - radius, ( *pointIt ).x() + radius, ( *pointIt ).y() + radius );
+
     // get the data
-    float *dataBuffer = ( float * ) CPLMalloc( sizeof( float ) * blockSize * blockSize );
-    if ( GDALRasterIO( mRasterBandH, GF_Read, xPosition, yPosition, blockSize, blockSize,
-                       dataBuffer, blockSize, blockSize, GDT_Float32, 0, 0 ) != CE_None )
-    {
-      result = RasterIoError;
-    }
+    QgsRasterBlock *block = mProvider->block( 1, extent, blockSize, blockSize );
+    QByteArray blockData = block->data();
+    float *dataBuffer = ( float * ) blockData.data();
```

Comment:
can this be changed to `static_cast` to be more cpp style?

Reply:
Hmm, I'm not a guru, but it looks like we can not use `static_cast` here without implementing custom conversion

Reply:
no worries, it was just a random thought. In general the new-style is more explicit about what's happening in the background.